Add parallel front end goal in 25h2 (#330)
rust-lang/rust-project-goals: src/2025h2/parallel-front-end.md

# Promoting Parallel Front End

| Metadata         |                                                   |
| :--------------- | ------------------------------------------------- |
| Point of contact | @SparrowLii                                       |
| Teams            | <!-- TEAMS WITH ASKS -->                          |
| Task owners      | <!-- TASK OWNERS -->                              |
| Status           | Proposed                                          |
| Tracking issue   | [https://github.com/rust-lang/rust/issues/113349] |
| Zulip channel    | [#t-compiler/wg-parallel-rustc][channel]          |

[channel]: https://rust-lang.zulipchat.com/#narrow/channel/187679-t-compiler.2Fwg-parallel-rustc/

## Summary

> Continue to parallelize front-end stabilization and performance improvements, continuing from the [2025h1 goal](https://rust-lang.github.io/rust-project-goals/2025h1/parallel-front-end.html).

Reviewer: ```suggestion
Continue with stabilization and performance improvements to parallel front-end, continuing from the [2025h1 goal](https://rust-lang.github.io/rust-project-goals/2025h1/parallel-front-end.html).
```

## Motivation

> Verify and resolve the few remaining deadlock and other issues, stabilize the features, and try to further improve the parallel compilation performance through several means.

Reviewer: ```suggestion
Verify and resolve the few remaining deadlock and other issues, stabilize the feature, and try to further improve parallel compilation performance through various means.
```

Author: Done, thansk!

### The status quo

> Parallel front end progressed well in the first half of the year. We resolved the deadlock issue caused by the deadlock handler, added the rustc-rayon  that the parallel front end depends on to the rustc working tree, and optimized the work-stealing scheduling algorithm to avoid deadlocks in the parallelization of rustc queries.

Reviewer: ```suggestion
Parallel front end progressed well in the first half of the year. We resolved the deadlock issue caused by the deadlock handler, added the rustc-rayon dependency of the parallel front end to the rustc working tree, and changed the work-stealing scheduling algorithm to avoid deadlocks in parallel execution of rustc queries.
```

> Next we need to resolve and verify the remaining issue list, improve the parallel front end test suit, and include all known issues in the ui tests to ensure the stability of the features.

Reviewer: ```suggestion
Next we need to resolve and verify the remaining issue list, improve the parallel front end test suite, and include all known issues into the ui tests to ensure stability of the feature.
```

Then we will start the feature stabilization process to push parallel front end into the stable version.

In addition, there are already contributors working on improving the performance of the parallel front end, such as trying to parallelize the macro expansion process, analysing data contention in query calls, etc. (thank you very much). We will conduct more detailed analysis and implementation of these directions in the second half of the year.

> In addition, we need to document and enhance the support of surrounding tools for parallel frontends, such as bootstrap, rustc-perf, and Cargo.

Reviewer: ```suggestion
In addition, we need to document and enhance the support of surrounding tools for the parallel frontend, such as bootstrap, rustc-perf, and Cargo.
```

Author: Done, thansk!

### The next 6 months

> - Resolve the remaining issues in the issue [list](https://github.com/rust-lang/rust/issues?q=state%3Aopen%20label%3AA-parallel-compiler). Most of them have been solved, but are lack of ways to ensure stability.

Reviewer: ```suggestion
- Resolve the remaining issues in the issue [list](https://github.com/rust-lang/rust/issues?q=state%3Aopen%20label%3AA-parallel-compiler). Most of them have been solved, but are missing tests ensuring that the code won't regress again.
```

> - Land the parallel front end test [suite](https://github.com/rust-lang/rust/pull/132051) to ensure the robustness and prevent various issues from occurring again.

Reviewer: ```suggestion
- Land the parallel front end [test suite](https://github.com/rust-lang/rust/pull/132051) to ensure the robustness and prevent various issues from occurring again.
```

- Enable parallel front end in bootstrap.

> - Continue to improve parallel compilation performance, via parallelize the macro expansion and reduce data competition.

Reviewer: ```suggestion
- Continue to improve parallel compilation performance, by parallelizing macro expansion and reducing data contention.
```

Author: Done, thansk!

- Enable parallel front end in Cargo.

> - Add more test cases for parallel front end in rustc-perf

Reviewer: ```suggestion
- Add more benchmarks for the parallel front end to rustc-perf
```

Author: Done, thansk!

> - Add more test cases for parallel front end in rustc-perf

Reviewer: Can add a couple more items
- Enable the parallel frontend by default on nightly, but not on stable, for preliminary testing (similarly to [LLD](https://github.com/rust-lang/rust/pull/140525)) 
- Write stabilization report for the feature and submit it for compiler FCP

Author: Thanks! I added the second one. 

And enable it by default (even in nightly) might require more statistics, as it may cause performance regressions compared to single-threaded in many cases.

### The "shiny future" we are working towards

> We will use measures to ensure the robustness of the parallel front end and push it into the stable version.

Reviewer: ```suggestion
We will ensure robustness of the parallel front end and push it to stabilization.
```

The current parallelization front end can already reduce the overall compilation time by 20~30+ percent, but we will continue to optimize it so that this number continues to grow.

## Design axioms

The parallel front end should be:
- safe: Ensure the safe and correct execution of the compilation process

> - consistent: The compilation result should be consistent with that in single thread

Reviewer: ```suggestion
- consistent: The compilation result should be consistent with that in single thread by default
```
I'm still interested in possible speedups in development scenarios through reducing reproducibility

Author: Done, thanks!

# Promoting Parallel Front End

| Metadata         |                                                   |
| :--------------- | ------------------------------------------------- |
| Point of contact | @SparrowLii                                       |
| Teams            | <!-- TEAMS WITH ASKS -->                          |
| Task owners      | <!-- TASK OWNERS -->                              |
| Status           | Proposed                                          |

> | Tracking issue   | [https://github.com/rust-lang/rust/issues/113349] |

Reviewer 2: Probably something like this

```suggestion
| Tracking issue   | [rust-lang/rust-project-goals#121]                |
| See also         | [rust-lang/rust#113349]                           |
```
